src/http/HttpClient/index.ts

```diff
@@ -169,6 +169,102 @@ export default class HttpClient implements IHttpClient {
+        url: string,
+        form: FormData,
+        onProgress?: (percentage: number, event: ProgressEvent<XMLHttpRequestEventTarget>) => void
+    ): Promise<HttpResponse<TResponse>> {
+        const token = await this.authContainer.acquireTokenAsync(url);
+        return new Promise((resolve, reject) => {
+            const xhr = new XMLHttpRequest();
+
+            if (onProgress) {
+                xhr.upload.addEventListener('progress', e => {
+                    if (e.lengthComputable) {
+                        const percentage = Math.round((e.loaded * 100) / e.total);
+                        onProgress(percentage, e);
+                    }
+                });
+            }
+
+            xhr.addEventListener('load', () => {
+                const headerLines = xhr.getAllResponseHeaders();
+                const headers = headerLines.trim().split(/[\r\n]+/);
+
+                var headerMap = new Headers();
+                headers.forEach(function(line) {
+                    const parts = line.split(': ');
+                    const header = parts.shift();
+                    const value = parts.join(': ');
+                    if (header) headerMap.append(header, value);
+                });
+
+                const response: HttpResponse<TResponse> = {
+                    data: JSON.parse<TResponse>(xhr.responseText),
+                    status: xhr.status,
+                    headers: headerMap,
+                    refreshRequest: null,
+                };
+                resolve(response);
+            });
+
+            xhr.upload.addEventListener('error', () => {
+                const response = xhr.responseText;
+                if (response) {
+                    const errorResponse = JSON.parse<TExpectedErrorResponse>(response);
+                    reject(new HttpClientRequestFailedError(url, xhr.status, errorResponse));
+                }
+                reject(new HttpClientRequestFailedError(url, xhr.status, null));
+            });
+
+            xhr.setRequestHeader('X-Session-Id', this.sessionId);
+            xhr.setRequestHeader('Accept', 'application/json');
+            xhr.setRequestHeader('x-pp-refresh', 'true');
+            xhr.setRequestHeader('Authorization', 'Bearer ' + token);
+
+            xhr.open('POST', url, true);
+
+            xhr.send(form);
+        });
+    }
+
+    async getBlobAsync<TExpectedErrorResponse>(
+        url: string,
+        init?: RequestInit | null
+    ): Promise<File> {
+        if (!init) {
+            init = {
+                headers: new Headers({ Accept: '*/*' }),
+            };
+        }
+
+        init = ensureRequestInit(init, init => ({
+            ...init,
+            method: 'GET',
+        }));
+
+        const response = await this.performFetchAsync<TExpectedErrorResponse>(url, init);
+        const blob = await response.blob();
+
+        const contentDisposition = response.headers.get('Content-Disposition');
+        let filename;

```

Comment:
Could this be done using Array.Prototype.Find ? Maybe something like this

 
```suggestion
    if (contentDisposition === null) {
        throw new Error('Cannot download file without filename');
    }
    const parts = contentDisposition.split(';');
    const fileNamePart = parts.find(part =>  part.indexOf('filename=') !== -1)
    const filename = fileNamePart.split('=')[1];


    return new File([blob], filename);
```

Doing it this way we avoid nested if's and variables

Reply:
The suggested change didn't work over multiple lines, so don't commit from here :D

Reply:
Good point. Extraxted the file name logic to a helper method.